lib/localenv/detect.go

```diff
@@ -0,0 +1,70 @@
+/*
+Copyright 2019 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package localenv
+
+import (
+	"context"
+
+	"github.com/gravitational/gravity/lib/defaults"
+
+	"github.com/gravitational/trace"
+)
+
+// DetectCluster attempts to detect whether there is a deployed cluster.
+//
+// The behavior is as follows:
+//
+//   - If the cluster is deployed/healthy, returns nil.
+//   - If the cluster is not deployed and the node's clean, returns NotFound.
+//   - If partial state is detected, returns a specific error.
+func DetectCluster(ctx context.Context, env *LocalEnvironment) error {
```

Comment:
- I believe this API belongs elsewhere (perhaps in lib/status)
- I don't think it will work as described (Initial `SiteOperator` call failure will not let you advance to look at packages). I would split it into functional parts:
```go
// DetectCluster attempts to detect whether there is a deployed cluster.
//
// The behavior is as follows:
//
//   - If the cluster is deployed/healthy, returns nil.
//   - If the cluster is not deployed and the node's clean, returns NotFound.
//   - If partial state is detected, returns a specific error.
func DetectCluster(ctx context.Context, env *LocalEnvironment) error {
	err := isClusterController(env)
	if err == nil {
		return nil
	}
	err := isPackageState(env.Packages)
	if err != nil {
		return trace.Wrap(err)
	}
	// There are local packages: assume partial installation state or
	// degraded state and log the original error for troubleshooting.
	log.WithError(getErr).Warn("Failed to query local cluster.")
	return trace.BadParameter(`Detected local Gravity state on the node but the cluster is inaccessible.
This usually means one of two things:

* The cluster is partially installed/uninstalled.
* The cluster is degraded.

To clean up the node from the partial state, run "gravity leave --force".
Otherwise, run "gravity status" and troubleshoot the degraded cluster.
`)
}

func isClusterController(env *localenv.LocalEnvironment) error {
	clusterOperator, err := env.SiteOperator()
	if err == nil {
		_, getErr := clusterOperator.GetLocalSite()
		if getErr == nil {
			return nil
		}
	}
	return trace.Wrap(err)
}

func isPackageState(packages pack.PackageService) error {
	// Check local packages to see if there is a local state.
	packages, err := packages.GetPackages(defaults.SystemAccountOrg)
	if err != nil {
		return trace.Wrap(err)
	}
	// There are no local packages: assume this is a clean node.
	if len(packages) == 0 {
		return trace.NotFound("no cluster installed")
	}
	return nil
}
```

Reply:
Can't put it into lib/status as-is due to cyclic deps (that's where I originally tried to place it) but I can make a function there that accepts operator/packages and call it from here.

Agree about splitting into functions.

Reply:
@a-palchikov I've split the function into parts but decided to keep it where it is due to circular deps issues I mentioned above as I don't think the effort of trying to split it further and move worth it at this point.

Reply:
Also, I tested all 3 scenarios - install checks on a clean node, upgrade checks when there's a cluster and error in the partial state case - and it works fine now.